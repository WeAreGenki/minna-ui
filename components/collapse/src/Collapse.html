<!--
  COLLAPSE COMPONENT

  TODO: Write basic info.

  USAGE:
    Example:

      <MinnaCollapse>
        <p>Your content.</p>
      </MinnaCollapse>
-->

<button class="button button-collapse" on:click="set({ isOpen: !isOpen })">
  {isOpen ? textClose : textOpen}
</button>

<div class="collapse {isOpen ? '' : 'collapse-hide'}"><slot></slot></div>

<script>
  export default {
    data: () => ({
      isOpen: false,

      /* optional props */
      textOpen: 'Expand ▾',
      textClose: 'Collapse ▴',
    }),
  };
</script>

<style>
  .collapse-hide {
    display: none;
  }
</style>
