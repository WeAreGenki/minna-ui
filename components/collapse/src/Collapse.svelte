<!--
  COLLAPSE COMPONENT

  TODO: Write basic info.

  USAGE:
    Example:

      <MinnaCollapse>
        <p>Your content.</p>
      </MinnaCollapse>

  @format
-->

<script>
  export let textOpen = 'Expand ▾';
  export let textClose = 'Collapse ▴';
  export let isOpen;

  function handleClick() {
    isOpen = !isOpen;
  }
</script>

<style>
  /* stylelint-disable a11y/no-display-none */
  .collapse-hide {
    display: none;
  }
</style>

<svelte:options tag="minna-collapse" />

<button class="button button-collapse" on:click="{handleClick}">
  {isOpen ? textClose : textOpen}
</button>

<div class="collapse {isOpen ? '' : 'collapse-hide'}">
  <slot></slot>
</div>
