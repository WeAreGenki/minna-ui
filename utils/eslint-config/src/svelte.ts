/**
 * Minna UI Svelte & Sapper ESLint config add-on.
 * @fileoverview Provides extra config for Svelte & Sapper projects. This
 * config is and add-on which should extend the base config preset.
 * @see https://eslint.org/docs/user-guide/configuring
 */

/* eslint-disable sort-keys */

interface BlockAttributes {
  type: string;
}

export = {
  plugins: ['svelte3'],
  settings: {
    'svelte3/ignore-styles': (attr: BlockAttributes) =>
      attr.type === 'text/postcss',
  },
  overrides: [
    // Svelte components
    {
      files: ['*.svelte'],
      rules: {
        // import order cannot be determined correctly in .svelte components
        'import/first': 'off',
        // components are compiled at build-time so devDependencies are OK
        'import/no-extraneous-dependencies': [
          'error',
          { devDependencies: true },
        ],
        // Svelte abuses the syntax with `export let`
        'import/no-mutable-exports': 'off',
      },
    },
  ],
};
