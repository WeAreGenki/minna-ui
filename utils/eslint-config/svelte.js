/**
 * Minna UI Svelte & Sapper ESLint config add-on.
 *
 * @fileoverview Provides extra config for Svelte & Sapper projects. This
 * config is and add-on which should extend the base config preset.
 * @see https://eslint.org/docs/user-guide/configuring
 */

/* eslint-disable sort-keys */

'use strict';

/** @type {import('./types').ESLintConfig} */
module.exports = {
  plugins: ['svelte3'],
  settings: {
    /**
     * @param {import('./types').BlockAttributes} attr - Tag attributes.
     * @returns {boolean} If style block should be ignored.
     */
    'svelte3/ignore-styles': (attr) => attr.type === 'text/postcss',
    // TODO: Uncomment once false positives are handled better
    // 'svelte3/lint-template': true,
  },
  overrides: [
    // Svelte components
    {
      files: ['*.svelte'],
      rules: {
        // import order cannot be determined correctly in .svelte components
        'import/first': 'off',
        // components are compiled at build-time so devDependencies are OK
        'import/no-extraneous-dependencies': [
          'error',
          { devDependencies: true },
        ],
        // Svelte abuses the syntax with `export let`
        'import/no-mutable-exports': 'off',
      },
    },
  ],
};
