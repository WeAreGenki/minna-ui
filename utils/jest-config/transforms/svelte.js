/**
 * Jest transform for compiling Svelte components into JavaScript.
 *
 * NOTE: `svelte` is not listed as a dependency so developers can use this
 * jest preset for non-svelte projects.
 */

'use strict';

const { compile } = require('svelte/compiler'); // eslint-disable-line import/no-extraneous-dependencies

/** @type {import('../types').JestTransformer} */
exports.process = (src, filename) => {
  // Strip out <style> tags to prevent errors when unable to parse PostCSS etc.
  const re = /<style[^>]*>[\S\s]*?<\/style>/g;
  const normalised = src.replace(re, '');

  const result = compile(normalised, {
    accessors: true, // Easy access to component props for tests
    css: false,
    dev: true,
    filename,
    format: 'cjs',
    preserveWhitespace: true,
  });

  const esInterop =
    'Object.defineProperty(exports, "__esModule", { value: true });';

  return {
    code: result.js.code + esInterop,
    map: result.js.map,
  };
};
