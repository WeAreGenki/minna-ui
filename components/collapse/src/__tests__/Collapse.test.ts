// TODO: Add integration/UI test to check the content is visually not shown

/* eslint-disable @typescript-eslint/no-non-null-assertion */

import { tick } from 'svelte';
import Collapse from '../Collapse.svelte';

describe('Collapse component', () => {
  it('renders correctly with no props', () => {
    expect.assertions(3);
    function wrapper(): void {
      const target = document.createElement('div');
      new Collapse({ target });
      expect(target.querySelector('.collapse-hide')).not.toBeNull();
      expect(target.innerHTML).toMatchSnapshot();
    }
    expect(wrapper).not.toThrow();
  });

  it('renders correctly with textOpen/textClose props', async () => {
    expect.assertions(4);
    const target = document.createElement('div');
    const component = new Collapse({
      props: {
        textClose: 'less',
        textOpen: 'more',
      },
      target,
    });
    const button = target.querySelector<HTMLButtonElement>('.button-collapse')!;
    expect(button.innerHTML).toEqual('more');
    expect(target.innerHTML).toMatchSnapshot();
    component.$set({ isOpen: true });
    await tick();
    expect(button.innerHTML).toEqual('less');
    expect(target.innerHTML).toMatchSnapshot();
  });

  it('is collapsed by default', () => {
    expect.assertions(2);
    const target = document.createElement('div');
    const component = new Collapse({ target });
    expect(component.$$.ctx.isOpen).toBeFalsy();
    expect(target.querySelector('.collapse-hide')).not.toBeNull();
  });

  it('shows collapsed content on click', async () => {
    expect.assertions(4);
    const target = document.createElement('div');
    const component = new Collapse({ target });
    const button = target.querySelector('button')!;
    expect(component.$$.ctx.isOpen).toBeFalsy();
    expect(target.querySelector('.collapse-hide')).not.toBeNull();
    button.click();
    await tick();
    expect(component.$$.ctx.isOpen).toBeTruthy();
    expect(target.querySelector('.collapse-hide')).toBeNull();
  });

  it('hides collapsed content on click', async () => {
    expect.assertions(4);
    const target = document.createElement('div');
    const component = new Collapse({
      props: {
        isOpen: true,
      },
      target,
    });
    const button = target.querySelector('button')!;
    expect(component.$$.ctx.isOpen).toBeTruthy();
    expect(target.querySelector('.collapse-hide')).toBeNull();
    button.click();
    await tick();
    expect(component.$$.ctx.isOpen).toBeFalsy();
    expect(target.querySelector('.collapse-hide')).not.toBeNull();
  });
});
