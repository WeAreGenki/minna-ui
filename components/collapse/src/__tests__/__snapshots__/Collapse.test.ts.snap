// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collapse component renders correctly with no props 1`] = `"<button class=\\"button button-collapse\\">Expand â–¾</button> <div class=\\"collapse collapse-hide\\"></div>"`;

exports[`Collapse component renders correctly with textOpen/textClose props 1`] = `"<button class=\\"button button-collapse\\">more</button> <div class=\\"collapse collapse-hide\\"></div>"`;

exports[`Collapse component renders correctly with textOpen/textClose props 2`] = `"<button class=\\"button button-collapse\\">less</button> <div class=\\"collapse \\"></div>"`;
